controllers/idle_handlers.go

```diff
@@ -18,103 +18,205 @@ package controllers
 	ctrl "sigs.k8s.io/controller-runtime"
 )
 
-func (r *ImageBasedUpgradeReconciler) cleanupUnbootedStateroot(stateroot string) error {
+//nolint:unparam
+func (r *ImageBasedUpgradeReconciler) handleAbort(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
+	r.Log.Info("Starting handleAbort")
+
+	if r.cleanup(ctx, false, ibu) {
+		r.Log.Info("Finished handleAbort")
+		return doNotRequeue(), nil
+	}
+
+	utils.SetStatusCondition(&ibu.Status.Conditions,
+		utils.ConditionTypes.Idle,
+		utils.ConditionReasons.AbortFailed,
+		metav1.ConditionFalse,
+		"Unable to cleanup successfully",
+		ibu.Generation,
+	)
+	return doNotRequeue(), nil
+}
+
+//nolint:unparam
+func (r *ImageBasedUpgradeReconciler) handleAbortFailure(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
+	return doNotRequeue(), nil
+}
+
+//nolint:unparam
+func (r *ImageBasedUpgradeReconciler) handleFinalize(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
+	r.Log.Info("Starting handleFinalize")
+
+	if r.cleanup(ctx, true, ibu) {
+		r.Log.Info("Finished handleFinalize")
+		return doNotRequeue(), nil
+	}
+	utils.SetStatusCondition(&ibu.Status.Conditions,
+		utils.ConditionTypes.Idle,
+		utils.ConditionReasons.FinalizeFailed,
+		metav1.ConditionFalse,
+		"Unable to cleanup successfully",
+		ibu.Generation,
+	)
+	return doNotRequeue(), nil
+}
+
+//nolint:unparam
+func (r *ImageBasedUpgradeReconciler) handleFinalizeFailure(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
+
+	// TODO actual steps
+	// If succeeds, return doNotRequeue
+	return doNotRequeue(), nil
+}
+
+// cleanup cleans stateroots, precache, backup, ibu files and OADP operator
+// returns true if all cleanup tasks were successful
+func (r *ImageBasedUpgradeReconciler) cleanup(
+	ctx context.Context, allUnbootedStateroots bool, ibu *lcav1alpha1.ImageBasedUpgrade) bool {
+	// try to clean up as much as possible and avoid returning when one of the cleanup tasks fails
+	// successful means that all the cleanup tasks completed without any error
+	successful := true
+
+	if err := r.cleanupStateroots(allUnbootedStateroots, ibu); err != nil {
+		successful = false
+		r.Log.Error(err, "failed to cleanup stateroots")
+	}
+	if err := r.Precache.Cleanup(ctx); err != nil {
+		successful = false
+		r.Log.Error(err, "failed to cleanup precaching resources")
+	}
+	if allRemoved, err := r.BackupRestore.CleanupBackups(ctx); err != nil {
+		successful = false
+		r.Log.Error(err, "failed to cleanup backups")
+	} else if !allRemoved {
+		successful = false
+		r.Log.Error(errors.New("failed to delete all the backup CRs"), "")
+	}
+	if err := r.BackupRestore.DeleteOadpOperator(ctx, backuprestore.OadpNs); err != nil {
+		successful = false
+		r.Log.Error(err, "failed to delete OADP operator")
+	}
+	if err := cleanupIBUFiles(); err != nil {
+		successful = false
+		r.Log.Error(err, "failed to cleanup ibu files")
```

Comment:
Maybe we can provide all those errors in status? User should not go to logs in order to see failures he can actually see in status? no?